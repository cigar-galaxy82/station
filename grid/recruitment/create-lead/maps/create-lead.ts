/* eslint-disable jest/no-export, jest/valid-describe, jest/valid-title, jest/no-identical-title */

import { RecordingProcessOptions, SuperfaceTest } from '@superfaceai/testing';
import { readFileSync } from 'fs';

import { buildSuperfaceTest } from '../../../test-config';

const sampleLead = {
  name: 'John Lead',
  firstName: 'John',
  lastName: 'Lead',

  email: 'john.lead@fakemail.com',
  phone: '1-859-557-6573',
  address: '25772 Gustave Shore, Iowa, USA',

  education: [
    {
      degree: 'MBA',
      school: 'University of Pennsylvania',
      fieldOfStudy: null,
      startedAt: '2008-03-01',
      endedAt: '2011-03-30',
    },
    {
      degree: 'B.S.',
      school: 'University of Chicago',
      fieldOfStudy: 'Marketing Communication & Economics',
      startedAt: '2004-09-01',
      endedAt: '2007-03-30',
    },
  ],

  workExperience: [
    {
      position: 'Sales Director',
      summary: 'Summary of a work experience at Test Company',
      company: 'Test Company',
      industry: 'Telecommunications',
      current: false,
      startedAt: '2011-03-01',
      endedAt: '2014-03-30',
    },
  ],

  cv: {
    name: 'cv-sample.pdf',
    data: readFileSync('./grid/recruitment/cv-sample.pdf', {
      encoding: 'base64',
    }),
  },

  links: [
    {
      name: 'twitter',
      url: 'https://twitter.com/candidate.username',
    },
    {
      name: 'Portfolio',
      url: 'https://url.to.portfolio',
    },
  ],

  source: ['LinkedIn', 'Indeed', 'Glassdoor'],
};

const describeIf = (
  condition: boolean,
  ...args: Parameters<jest.Describe>
): void => (condition ? describe(...args) : describe.skip(...args));

export const createLeadTest = (
  provider: string,
  jobIds: { valid: string; invalid: string },
  options?: RecordingProcessOptions
): void => {
  describe(`recruitment/create-lead/${provider}`, () => {
    let superface: SuperfaceTest;

    beforeEach(() => {
      superface = buildSuperfaceTest({
        profile: 'recruitment/create-lead',
        provider,
      });
    });

    describe('CreateLead', () => {
      describe('when specified job does exist', () => {
        it('should perform successfully', async () => {
          const result = await superface.run(
            {
              useCase: 'CreateLead',
              input: {
                jobId: jobIds.valid,
                ...sampleLead,
              },
            },
            options
          );

          expect(() => result.unwrap()).not.toThrow();
          expect(result).toMatchSnapshot();
        });
      });

      describe('when specified job does not exist', () => {
        it('should map error', async () => {
          const result = await superface.run(
            {
              useCase: 'CreateLead',
              input: {
                jobId: jobIds.invalid,
                firstName: 'Demo',
                lastName: 'Testing',
                email: 'demo_testing@fakemail.com',
              },
            },
            options
          );

          expect(() => result.unwrap()).toThrow();
          expect(result).toMatchSnapshot();
        });
      });

      describeIf(
        provider === 'breezy-hr',
        'when specified company does not exist',
        () => {
          let companyId: string;

          beforeAll(() => {
            companyId = process.env.BREEZY_HR_COMPANY_ID!;

            process.env.BREEZY_HR_COMPANY_ID = '1b111c1111ef11';
          });

          afterAll(() => {
            process.env.BREEZY_HR_COMPANY_ID = companyId;
          });

          it('should map error', async () => {
            const result = await superface.run(
              {
                useCase: 'CreateLead',
                input: {
                  jobId: jobIds.valid,
                  firstName: 'Demo',
                  lastName: 'Testing',
                  email: 'demo_testing@fakemail.com',
                },
              },
              options
            );

            expect(() => result.unwrap()).toThrow();
            expect(result).toMatchSnapshot();
          });
        }
      );

      describeIf(
        provider === 'workable',
        'when specified subdomain does not exist',
        () => {
          let subdomain: string;

          beforeAll(() => {
            subdomain = process.env.WORKABLE_SUBDOMAIN!;

            process.env.WORKABLE_SUBDOMAIN = 'invalid-superface';
          });

          afterAll(() => {
            process.env.WORKABLE_SUBDOMAIN = subdomain;
          });

          it('returns error', async () => {
            const result = await superface.run(
              {
                useCase: 'CreateLead',
                input: {
                  jobId: jobIds.valid,
                  firstName: 'Demo',
                  lastName: 'Testing',
                  email: 'demo_testing@fakemail.com',
                },
              },
              options
            );

            expect(() => result.unwrap()).toThrow();
            expect(result).toMatchSnapshot();
          });
        }
      );
    });
  });
};
