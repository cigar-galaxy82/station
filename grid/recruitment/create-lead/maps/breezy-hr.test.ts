import { createLeadTest } from './create-lead';

createLeadTest('breezy-hr', {
  valid: 'f1cfc5572a5701',
  invalid: '12345ab789d1dd',
});
