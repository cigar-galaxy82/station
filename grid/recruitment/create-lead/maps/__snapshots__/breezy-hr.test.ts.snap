// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recruitment/create-lead/breezy-hr CreateLead should map error 1`] = `
Err {
  "error": "MappedError: Expected error
Properties: {
  \\"title\\": \\"Unknown error\\",
  \\"detail\\": {
    \\"type\\": \\"addCandidateError\\",
    \\"message\\": \\"there was an unknown error creating candidate.\\"
  },
  \\"code\\": \\"UnknownError\\",
  \\"rateLimit\\": {
    \\"bucket\\": \\"breezy-hr\\",
    \\"totalRequests\\": 100,
    \\"remainingRequests\\": 96,
    \\"remainingRequestsPercentage\\": 96,
    \\"resetTimestamp\\": 1671111168
  }
}
AST Path: definitions[0].statements[2].statements[0]
Original Map Location: Line 24, column 5",
}
`;

exports[`recruitment/create-lead/breezy-hr CreateLead should perform successfully 1`] = `
Ok {
  "value": Object {
    "id": "LEAD:c82c6b873b5701",
    "jobId": "f1cfc5572a5701",
    "rateLimit": Object {
      "bucket": "breezy-hr",
      "remainingRequests": 98,
      "remainingRequestsPercentage": 98,
      "resetTimestamp": 1671111168,
      "totalRequests": 100,
    },
  },
}
`;
